File: Algorithms/Math/0008_myAtoi.c
// Practice for: 0008_myAtoi.py
// Bridge logic from Python to C pointers here.

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int myAtoi (char *s){
    //if the first instance is a non-digit, return zero
    int i = 0;
    int sign = 1; //default
    long res = 0;

    while (s[i] == ' '){ //traverse whitspaces 
        i++;
    }
    //sign
    if (s[i] == '-'){
        sign = -1;
        i++;
    } else if (s[i] == '+'){
        sign = 1;
        i++;
    }

    while (s[i] >= '0' && s[i] <= '9'){
        res = res * 10 + (s[i] - '0');
        i++;
    }

    //check bounds
    if (res * sign <= INT_MIN){
        return INT_MIN;
    }
    if (res * sign >= INT_MAX){
        return INT_MAX;
    }

    return res*sign;
}

int main() {
    char* test_strings[] = {
        "42", 
        "   -42", 
        "4193 with words", 
        "words and 987", 
        "91283472332", 
        "-91283472332",
        "+1"
    };

    int expected[] = {
        42, 
        -42, 
        4193, 
        0, 
        INT_MAX, 
        INT_MIN, 
        1
    };

    printf("--- Running myAtoi C Tests ---\n");
    for (int i = 0; i < 7; i++) {
        int result = myAtoi(test_strings[i]);
        if (result == expected[i]) {
            printf("PASS: [%s] -> %d\n", test_strings[i], result);
        } else {
            printf("FAIL: [%s] -> Expected %d, but got %d\n", 
                   test_strings[i], expected[i], result);
        }
    }
    return 0;
}

File: Algorithms/Math/0008_myAtoi.py
#math 

class Solution:
    def myAtoi(self, s: str) -> int:
        if not s:
            return 0

        INT_MAX = 2**31 - 1
        INT_MIN = -2**31

        i = 0
        while i < len(s) and s[i] == ' ': #traverse whitespaces 
            i+=1

        if i == len(s): #all whitespaces
            return 0

        sign = 1
        if s[i] == '-': 
            sign = -1
            i+=1 #traverse sign
        elif s[i] == '+': 
            sign = 1
            i+=1 #traverse sign

        res = 0
        while i < len(s) and s[i].isdigit():
            digit = int(s[i])
            res = res * 10 + digit

            #constraints 
            if sign*res <= INT_MIN:
                return INT_MIN
            if sign*res >= INT_MAX:
                return INT_MAX

            i+=1 #traverse 
        
        return res * sign

def test_atoi():
    sol = Solution()
    
    # 1. Standard Case
    assert sol.myAtoi("42") == 42
    
    # 2. Leading Whitespace & Negative Sign
    assert sol.myAtoi("   -42") == -42
    
    # 3. Reading Digits with Non-Digit Suffix
    assert sol.myAtoi("4193 with words") == 4193
    
    # 4. Words First (Should return 0)
    assert sol.myAtoi("words and 987") == 0
    
    # 5. Out of Range (Positive Overflow)
    assert sol.myAtoi("91283472332") == 2147483647 # INT_MAX
    
    # 6. Out of Range (Negative Overflow)
    assert sol.myAtoi("-91283472332") == -2147483648 # INT_MIN

    # 7. Edge Case: Just a sign
    assert sol.myAtoi("-") == 0

    print("✅ All Python test cases passed!")

if __name__ == "__main__":
    test_atoi()
